arch/common/shared_irq.c

```diff
@@ -22,3 +23,75 @@ void z_shared_isr(const void *data)
 		}
 	}
 }
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+
+static struct k_spinlock lock;
+
+void z_isr_install(unsigned int irq, void (*routine)(const void *),
+		   const void *param)
+{
+	struct z_shared_isr_table_entry *shared_entry;
+	struct _isr_table_entry *entry;
+	struct z_shared_isr_client *client;
+	unsigned int table_idx;
+	int i;
+	k_spinlock_key_t key;
+
+	table_idx = z_get_sw_isr_table_idx(irq);
+
+	/* check for out of bounds table index */
+	if (table_idx >= CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	shared_entry = &z_shared_sw_isr_table[table_idx];
+	entry = &_sw_isr_table[table_idx];
+
+	key = k_spin_lock(&lock);
+
+	/* have we reached the client limit? */
+	__ASSERT(shared_entry->client_num < CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS,
+		 "reached maximum number of clients");
+
+	if (entry->isr == z_irq_spurious) {
+		/* this is the first time a ISR/arg pair is registered
+		 * for INTID => no need to share it.
+		 */
+		entry->isr = routine;
+		entry->arg = param;
+
+		k_spin_unlock(&lock, key);
+
+		return;
+	} else if (entry->isr != z_shared_isr) {
+		/* INTID is being used by another ISR/arg pair.
+		 * Push back the ISR/arg pair registered in _sw_isr_table
+		 * to the list of clients and hijack the pair stored in
+		 * _sw_isr_table with our own shared_isr/irq_data pair.
+		 */
+		shared_entry->clients[shared_entry->client_num].isr = entry->isr;
+		shared_entry->clients[shared_entry->client_num].arg = entry->arg;
+
+		shared_entry->client_num++;
```

Comment:
Would not this potentially put you over CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS?

Reply:
This can only happen if `CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS` is 1, which I'm not sure why you'd do this. Either way, will add a `BUILD_ASSERT()` to enforce the fact that `CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS` should be >= 2. Thanks for pointing this out!